homeassistant/components/homematicip_cloud/config_flow.py

```diff
@@ -0,0 +1,106 @@
+"""Config flow to configure HomematicIP Cloud."""
+import asyncio
+import json
+
+import voluptuous as vol
+
+from homeassistant import config_entries, data_entry_flow
+from homeassistant.helpers import aiohttp_client
+
+from .const import DOMAIN, _LOGGER, CONF_ACCESSPOINT, CONF_PIN, CONF_NAME
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class HomematicipCloudFlowHandler(data_entry_flow.FlowHandler):
+    """Config flow HomematicIP Cloud."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize HomematicIP Cloud config flow."""
+        self.hmip_auth = None
+        self.hmip_apid = None
+        self.hmip_name = None
+
+    async def async_step_init(self, user_input=None):
+        """Handle a flow start."""
+        from homematicip.async.auth import AsyncAuth
+        from homematicip.base.base_connection import HmipConnectionError
+
+        errors = {}
+
+        if user_input is not None:
+            if user_input[CONF_ACCESSPOINT]:
+                self.hmip_apid = user_input[CONF_ACCESSPOINT]
+                self.hmip_name = user_input[CONF_NAME]
+                _LOGGER.info("Create new authtoken for %s", self.hmip_apid)
+
+                # Create new authtoken for the accesspoint
+                websession = aiohttp_client.async_get_clientsession(self.hass)
+                self.hmip_auth = AsyncAuth(self.hass.loop, websession)
+                try:
+                    await self.hmip_auth.init(self.hmip_apid)
+                except HmipConnectionError:
+                    return self.async_abort(reason='conection_aborted')
+                if user_input[CONF_PIN]:
+                    self.hmip_auth.pin = user_input[CONF_PIN]
+                try:
+                    state = await self.hmip_auth.connectionRequest(
+                        'HomeAssistant')
+                except HmipConnectionError:
+                    if state['errorCode'] == 'INVALID_PIN':
+                        errors[CONF_PIN] = 'invalid_pin'
+                    else:
+                        errors['base'] = 'register_failed'
+                else:
+                    # Connection established
+                    _LOGGER.info("Connection established")
+                    return await self.async_step_link()
+
+        return self.async_show_form(
+            step_id='init',
+            data_schema=vol.Schema({
+                vol.Required(CONF_ACCESSPOINT): str,
+                vol.Optional(CONF_PIN): str,
+                vol.Optional(CONF_NAME): str,
+            }),
+            errors=errors
+        )
+
+    async def async_step_link(self, user_input=None):
+        """Attempt to link with the HomematicIP Cloud accesspoint."""
+        from homematicip.base.base_connection import HmipConnectionError
+
+        errors = {}
+
+        _LOGGER.info("Wait for hardware button acknowledg")
+        # Wait for blue button pressed
+        wait = 30
+        state = False
+        while (not state and wait >= 0):
```

Comment:
I suggest you use the hue approach instead. Just check once, catch errors, then show form to user. When user confirms having pressed the button, this method will be called again and another check will be done and so on.

Reply:
Hi Martin, I would prefer the implemented approach:

- Init the connection with the access point ID
- move to link state and show the linking picture 
- wait there until the hardware button has been pressed
- if pressed finalise things and close the window else wait for timeout

This is how also the HomematicIP app is doing the linking process

Reply:
You don't need to use the while loop to wait until the button is pushed that is my point. Home assistant will call this method again when the user confirms having pushed the button in the frontend.

Reply:
@MartinHjelmare The user experience is better if he just have to press the hardware button and not need to pressed the hardware button & submit :-)

Reply:
🤷‍♂️

I don't like this construct.

Anyway, right now there's a min 30 second timeout until the form. Is shown to the user to press the button. You probably want to move something around.

Reply:
@MartinHjelmare For the construct I fully agree - its ugly (the problem with the 30sec form issue I fixed already), but I will change it to the hue variant. Maybe we find a better way later. I will commit this together with the fix fo the async package name which I need fix first. Thanks for your time!